runtime-modules/content/src/lib.rs

```diff
@@ -1435,227 +1453,320 @@ decl_module! {
                 text_hash,
-                forum_user_id,
+                member_id,
+                account_id,
+                first_post_init_bloat_bond,
+                params.post_mutable,
             );
 
             // Update next thread id
          <NextThreadId<T>>::mutate(|n| *n += One::one());
 
-         // increase category's thread counter
-         <CategoryById<T>>::mutate(category_id, |category| category.num_direct_threads += 1);
-
             // Generate event
             Self::deposit_event(
                 RawEvent::ThreadCreated(
                     new_thread_id,
-                    forum_user_id,
-                    category_id,
-                    title_hash,
-            channel_id,
+                    member_id,
+                    params,
                 )
             );
 
             Ok(())
      }
 
-    #[weight = 10_000_000]
-    fn delete_thread(
+       #[weight = 10_000_000]
+       fn delete_thread(
             origin,
-            forum_user_id: ForumUserId<T>,
-            category_id: T::CategoryId,
+            actor: ContentActor<T::CuratorGroupId, T::CuratorId, T::MemberId>,
             thread_id: T::ThreadId,
         ) -> DispatchResult {
             // Ensure data migration is done
-            // Self::ensure_data_migration_done()?;
+           // Self::ensure_data_migration_done()?;
+
+           let account_id = ensure_signed(origin)?;
+           let thread = Self::ensure_thread_exists(&thread_id)?;
+           let channel_id = thread.channel_id;
 
-            let account_id = ensure_signed(origin)?;
 
             Self::ensure_can_delete_thread(
                 &account_id,
-                &forum_user_id,
-                &category_id,
-                &thread_id
+                &actor,
+                &channel_id,
+                &thread_id,
             )?;
 
-            //
-            // == MUTATION SAFE ==
-            //
+           // iterate over values that share the first key
+            let iter = <PostById<T>>::iter_prefix_values(thread_id);
 
-            // Pay off to thread deleter
-//            Self::pay_off(thread_id, thread.cleanup_pay_off, &account_id)?;
+           // Pay off to author or burn tokens
+           if let ContentActor::Member(member) = actor {
+               if member == thread.author_id {
+                   let _ = Self::pay_off(thread_id, thread.bloat_bond, &account_id);
+               }
+           } else {
+                   let _ = balances::Module::<T>::burn(thread.bloat_bond);
+           }
 
-            // Delete thread
-            <ThreadById<T>>::remove(category_id, thread_id);
+           //
+           // == MUTATION SAFE ==
+           //
 
-            // decrease category's thread counter
-            <CategoryById<T>>::mutate(category_id, |category| category.num_direct_threads -= 1);
+           // burn all the other bloat bonds
+           let _ = iter.map(|post| balances::Module::<T>::burn(post.bloat_bond));
```

Comment:
A minor performance improvement you may want to keep in mind is to sometime try to aggregate payments, as each payment is a database hit (read and write, independently), which will add to the cost of the processing. We should not severely sacrifice clarity for this kind of optimization, but sometimes it can make sense.